### PR TITLE
[sdk 330] Implement consent request and send it to server.

### PR DESCRIPTION
**Pull request by ZahidZafar**

### CODE REVIEW

File: Assets/Plugins/CountlySDK/Services/InitializationCountlyService.cs [hunk 1]
```diff
@@ -20,6 +21,7 @@ internal InitializationCountlyService(CountlyConfiguration configuration, Countl
         internal async Task OnInitialisationComplete()
         {
             await StartSession();
+           _consentService.SendConsentChanges(_consentService._countlyConsents.Keys.ToList(), true);
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Services/InitializationCountlyService.cs:24]: This should not be here. Initial consent state should be sent after the SDK has initialised.

Comment (ZahidZafar) [reply Assets/Plugins/CountlySDK/Services/InitializationCountlyService.cs:24]: OnInitialisationComplete get after initialization of sdk.

File: Assets/Plugins/CountlySDK/Services/ConsentCountlyService.cs [hunk 2]
```diff
@@ -236,6 +273,10 @@ private void SetConsentInternal(Consents[] consents, bool value)
             }
 
             NotifyListeners(updatedConsents, value);
+            if (_sendConsentOnChange) {
```
[the remark this answers is not in the document]
Comment (ZahidZafar) [reply Assets/Plugins/CountlySDK/Services/ConsentCountlyService.cs:276]: first consent change should sent after sdk init. this check make sure it.

File: Assets/Plugins/CountlySDK/Services/InitializationCountlyService.cs (hunk 1 above)
Comment (ArtursKadikis) [reply Assets/Plugins/CountlySDK/Services/InitializationCountlyService.cs:24]: Mkey.
Consent changes should be sent before starting a session

File: Assets/Plugins/CountlySDK/Services/ConsentCountlyService.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArtursKadikis) [reply Assets/Plugins/CountlySDK/Services/ConsentCountlyService.cs:276]: Seems like a error prone way to do it

[the remark this answers is not in the document]
Comment (ZahidZafar) [reply Assets/Plugins/CountlySDK/Services/ConsentCountlyService.cs:276]: it's not, 
has been checked many times.

[the remark this answers is not in the document]
Comment (ArtursKadikis) [reply Assets/Plugins/CountlySDK/Services/ConsentCountlyService.cs:276]: Checking someting "multiple times" does not make it less error prone

[the remark this answers is not in the document]
Comment (ZahidZafar) [reply Assets/Plugins/CountlySDK/Services/ConsentCountlyService.cs:276]: it's not about checking, it makes sure first consent changes should send after init sdk,
Consent updates get calculate in'ConsentServcie' constructor.

[the remark this answers is not in the document]
Comment (ArtursKadikis) [reply Assets/Plugins/CountlySDK/Services/ConsentCountlyService.cs:276]: I was talking about your response

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Having all of these string operations every time is too wasteful performance and memory wise. Reacreate consent in a way that does not use any "ToLower" or "Replace" calls. Add also tests that validate that the resulting generated consent requests are in the correct format and the consent names have their names in the correct format.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Rename the starRating and remoteConfig enum names back to their previous values.

Add a few tests that verify that the consent request is created correctly.

**Comment by ArtursKadikis** [comment]
Additional tests should be added to verify that no consent request is sent when consent is not required.

File: Assets/Tests/PlayModeTests/ConsentTests.cs [hunk 3]
```diff
@@ -55,6 +58,88 @@ public void TestDefaultStateOfConsents()
...
+            Countly.Instance.Init(configuration);
+
+            Assert.IsNotNull(Countly.Instance.Consents);
+            Assert.AreEqual(1, Countly.Instance.CrashReports._requestCountlyHelper._requestRepo.Count);
+
+            CountlyRequestModel requestModel = Countly.Instance.Consents._requestCountlyHelper._requestRepo.Dequeue();
+            string myUri = requestModel.RequestUrl;
+            string consents = HttpUtility.ParseQueryString(myUri).Get("consent");
+            JObject json = JObject.Parse(consents);
+            Assert.IsTrue(json.GetValue("push").ToObject<bool>());
```
Comment (ArtursKadikis) [inline Assets/Tests/PlayModeTests/ConsentTests.cs:113]: Add an assert that validates how many keys are set in that json. Just to be sure that no others are there.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Consent values should be sent only if they change. Giving consent to a feature, that already has it, shouldn't produce a consent request.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Missed requested changes:
"Consent values should be sent only if they change. Giving consent to a feature, that already has it, shouldn't produce a consent request."

**Comment by ZahidZafar** [comment]
> Missed requested changes:
> "Consent values should be sent only if they change. Giving consent to a feature, that already has it, shouldn't produce a consent request."

there are already these test, we added before adding consent Rquest,

**Comment by ArtursKadikis** [comment]
What are you talking about? The current PR has a bug where 2 consent requests would be created if consent is given for the same feature twice in a row.

**Review by ArtursKadikis — APPROVED** [review]

**Pull request merged**